# Hands-on Exercise for  FPM Module

### 1. Exploring properties of the dataset accidents_10k.dat. Read more about it here:  http://fimi.uantwerpen.be/data/accidents.pdf

In [1]:
!head accidents_10k.dat

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 
2 5 7 8 9 10 12 13 14 15 16 17 18 20 22 23 24 25 27 28 29 32 33 34 35 36 37 38 39 
7 10 12 13 14 15 16 17 18 20 25 28 29 30 33 40 41 42 43 44 45 46 47 48 49 50 51 52 
1 5 8 10 12 14 15 16 17 18 19 20 21 22 24 25 26 27 28 29 30 31 41 43 46 48 49 51 52 53 54 55 56 57 58 59 60 61 
5 8 10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 31 33 36 38 39 41 43 46 56 62 63 64 65 66 67 68 
7 8 10 12 17 18 21 23 24 26 27 28 29 30 33 34 35 36 38 41 43 47 59 63 66 69 70 71 72 73 74 75 76 77 78 79 
1 12 14 15 16 17 18 21 22 23 24 25 27 28 29 30 31 35 38 41 43 44 53 56 57 58 59 60 63 66 80 81 82 83 84 
10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 30 31 33 39 41 43 44 46 49 59 60 62 63 66 82 
1 8 10 12 14 15 16 17 18 21 22 23 24 25 27 29 30 31 38 41 43 53 56 59 61 63 66 68 85 86 87 88 89 
1 8 12 13 14 15 16 17 18 22 24 25 28 30 38 41 42 43 46 49 60 63 64 66 80 82 84 90 91 92 93 94 95 


<span style="color:red">**Question 1a:** </span>. How many items are there in the data?

In [2]:
!awk -- '{for (i = 1; i <= NF; i++) wc[$i] += 1}; END {print length(wc)}' accidents_10k.dat

310


<span style="color:green">**Answer:** </span> Number of items in the data file accidents_10k are - 310

<span style="color:red">**Question 1b:** </span> How many transactions are present in the data?

In [3]:
!wc -l accidents_10k.dat

10000 accidents_10k.dat


<span style="color:green">**Answer:** </span> Number of transactions in the data file accidents_10k are - 10000

<span style="color:red">**Question 1c:** </span>.  What is the length of the smallest transaction?

In [11]:
!awk '{print NF}' accidents_10k.dat | sort -n | head -1

23


<span style="color:green">**Answer:** </span> Lenth of smallest transaction in data file accidents_10k.dat is - 23

<span style="color:red">**Question 1d:** </span>  What is the length of the longest transaction?

In [12]:
!awk '{print NF}' accidents_10k.dat | sort -n | tail -1

45


<span style="color:green">**Answer:** </span> Lenth of longest transaction in data file accidents_10k.dat is - 45

<span style="color:red">**Question 1e:** </span>  What is the size of the search space of frequent itemsets in this data?

In [15]:
2**310 - 1

2085924839766513752338888384931203236916703635113918720651407820138886450957656787131798913023

<span style="color:green">**Answer:** </span> The total possible search space is 2<sup>310</sup> − 1, but the actual frequent itemset search space is much smaller due to pruning.

<span style="color:red">**Question 1f:** </span>
Assume that you work for the deparment of transportation that collected this data. What benefit do you see in using itemset mining approaches on this data?

<span style="color:green">**Answer:** </span> I think, itemset mining can help us find combinations of conditions where accidents happen often, like maybe bad weather and high speed together. By seeing these frequent patterns, we can understand what situations are more risky. Then the department can take actions like improving roads, adding signs, or warning drivers to help prevent future accidents.

<span style="color:red">**Question 1g:** </span>  What type of itemsets (frequent, maximial or closed) would you be interested in discovering this dataset? State your reason.

<span style="color:green">**Answer:** </span> I would choose closed itemsets. Frequent itemsets can be too many and contain a lot of redundant information. Closed itemsets keep the same support information but reduce the number of patterns. This makes them easier to analyze while still keeping all important information about frequent patterns.

<span style="color:red">**Question 1h:** </span>  What minsup threshold would you use and why?

In [27]:
!N=$(wc -l < accidents_10k.dat); \
echo "N=$N"; \
echo "1% minsup =" $(( (N + 99) / 100 )); \
echo "0.5% minsup =" $(( (N + 199) / 200 )); \
echo "2% minsup =" $(( (N + 49) / 50 ))

N=10000
1% minsup = 100
0.5% minsup = 50
2% minsup = 200


<span style="color:green">**Answer:** </span> I would choose minsup around 100 (1% of data). If it is too small, there will be too many patterns. If it is too large, we may miss useful patterns. So a middle value is better.

### 2. Generating frequent, maximal and closed itemsets using $\color{red}{\text{Apriori}}$, $\color{red}{\text{ECLAT}}$, and $\color{red}{\text{FPGrowth}}$ algorihtms from the dataset accidents_10k.dat

<span style="color:red">**Question 2a:** </span> Generate frequent itemsets using Apriori, for minsup = 2000, 3000, and 4000. Which of these minsup thresholds results in a maximum number of frequent itemsets? Which of these minsup thresholds results in a least number of frequent itemsets? Provide a rationale for these observations.

In [28]:
!chmod u+x apriori

In [29]:
!./apriori

usage: ./apriori [options] infile [outfile]
find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
-t#      target type                              (default: s)
         (s: frequent, c: closed, m: maximal item sets,
          g: generators, r: association rules)
-m#      minimum number of items per set/rule     (default: 1)
-n#      maximum number of items per set/rule     (default: no limit)
-s#      minimum support of an item set/rule      (default: 10%)
-S#      maximum support of an item set/rule      (default: 100%)
         (positive: percentage, negative: absolute number)
-o       use original rule support definition     (body & head)
-c#      minimum confidence of an assoc. rule     (default: 80%)
-e#      additional evaluation measure            (default: none)
-a#      aggregation mode for evaluation measure  (default: none)
-d#      threshold for add. evaluation measure    (default: 10%)
-i       invalidate eva

In [30]:
!./apriori -ts -s-2000 accidents_10k.dat accidents_10k_Freq_S2000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [27.14s].
writing accidents_10k_Freq_S2000.txt ... [851034 set(s)] done [0.11s].


In [31]:
!./apriori -ts -s-3000 accidents_10k.dat accidents_10k_Freq_S3000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [38 item(s)] done [0.00s].
sorting and reducing transactions ... [9674/10000 transaction(s)] done [0.01s].
building transaction tree ... [24741 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 done [5.14s].
writing accidents_10k_Freq_S3000.txt ... [133799 set(s)] done [0.02s].


In [32]:
!./apriori -ts -s-4000 accidents_10k.dat accidents_10k_Freq_S4000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.43s].
writing accidents_10k_Freq_S4000.txt ... [29501 set(s)] done [0.00s].


<span style="color:green">**Answer:** </span> The maximum number of frequent itemsets (851034) occurs at minsup = 2000, and the least (29501) at minsup = 4000. This happens because a lower minsup allows more itemsets to be considered frequent, while a higher minsup is stricter and filters out many itemsets, leaving fewer results.

<span style="color:red">**Question 2b:** </span>   Using Apriori, compare the execution time for finding frequent itemsets for minsup = 2000, 3000, and 4000. Which of these minsup thresholds takes the least amount of time? Provide a rationale for this observation.

In [33]:
import datetime
start = datetime.datetime.now()
!./apriori -ts -s-2000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [26.66s].
writing <null> ... [851034 set(s)] done [0.02s].
26 secs  927187 microsecs


In [34]:
import datetime
start = datetime.datetime.now()
!./apriori -ts -s-3000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [38 item(s)] done [0.00s].
sorting and reducing transactions ... [9674/10000 transaction(s)] done [0.01s].
building transaction tree ... [24741 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 done [6.12s].
writing <null> ... [133799 set(s)] done [0.00s].
6 secs  341965 microsecs


In [35]:
import datetime
start = datetime.datetime.now()
!./apriori -ts -s-4000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.44s].
writing <null> ... [29501 set(s)] done [0.00s].
1 secs  509997 microsecs


<span style="color:green">**Answer:** </span> Least time is minsup = 4000 (about 1.5 seconds). With a higher minsup, fewer items are frequent, so Apriori generates far fewer candidate itemsets at each level.

<span style="color:red">**Question 2c:** </span> Using Apriori, find the frequent itemsets for minsup = 2000, 3000, and 4000. Determine the number of itemsets for each size (1 to max length of an itemset). What trends do you see that are common for all three minsup thresholds? What trends do you see that are different? Provide a rationale for these observations.

In [36]:
!awk '{print NF-1}' accidents_10k_Freq_S2000.txt|sort -n|uniq -c

     49 1
    705 2
   5285 3
  23745 4
  69647 5
 139628 6
 195730 7
 193299 8
 133819 9
  63937 10
  20497 11
   4189 12
    483 13
     21 14


In [37]:
!awk '{print NF-1}' accidents_10k_Freq_S3000.txt|sort -n|uniq -c

     38 1
    468 2
   2830 3
   9887 4
  21779 5
  31964 6
  32020 7
  21862 8
   9839 9
   2705 10
    387 11
     20 12


In [38]:
!awk '{print NF-1}' accidents_10k_Freq_S4000.txt|sort -n|uniq -c

     33 1
    319 2
   1492 3
   4043 4
   6926 5
   7751 6
   5626 7
   2546 8
    668 9
     91 10
      6 11


<span style="color:green">**Answer:** </span> In all three minsup values, the number of itemsets first increases with size, reaches a peak, and then decreases, forming a bell-shaped pattern. This happens because medium-sized itemsets have many possible combinations, while very large itemsets are less frequent and get pruned.

The difference is that when minsup increases, the total number of itemsets decreases, the peak shifts to smaller sizes, and the maximum length of itemsets becomes smaller. This is because a higher minsup is more strict, so fewer itemsets satisfy the support condition.


<span style="color:red">**Question 2d:** </span>  Using Apriori with minsup=2000, compare the number of frequent, maximal, and closed itemsets. Which is the largest set and which is the smallest set? Provide a rationale for these observations.

In [39]:
!./apriori -ts -s-2000 accidents_10k.dat accidents_10k_Freq_S2000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [26.99s].
writing accidents_10k_Freq_S2000.txt ... [851034 set(s)] done [0.10s].


In [40]:
!./apriori -tm -s-2000 accidents_10k.dat accidents_10k_Maximal_S2000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 14 done [44.70s].
filtering for maximal item sets ... done [0.06s].
writing accidents_10k_Maximal_S2000.txt ... [12330 set(s)] done [0.03s].


In [41]:
!./apriori -tc -s-2000 accidents_10k.dat accidents_10k_Closed_S2000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.03s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 14 done [44.95s].
filtering for closed item sets ... done [0.54s].
writing accidents_10k_Closed_S2000.txt ... [519902 set(s)] done [0.09s].


<span style="color:green">**Answer:** </span> The largest set is frequent itemsets (851034), and the smallest set is maximal itemsets (12330), with closed itemsets in between (519902). This is because frequent itemsets include all patterns, closed itemsets remove redundant ones with the same support, and maximal itemsets keep only the largest patterns.

<span style="color:red">**Question 2e:** </span> For a minsup = 2000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [45]:
!chmod u+x eclat
!chmod u+x fpgrowth

In [42]:
import datetime
start = datetime.datetime.now()
!./apriori -ts -s-2000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [28.68s].
writing <null> ... [851034 set(s)] done [0.02s].
29 secs  149892 microsecs


In [46]:
import datetime
start = datetime.datetime.now()
!./eclat -ts -s-2000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./eclat - find frequent item sets with the eclat algorithm
version 5.20 (2017.05.30)        (c) 2002-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
writing <null> ... [851034 set(s)] done [0.31s].
0 secs  411157 microsecs


In [47]:
import datetime
start = datetime.datetime.now()
!./fpgrowth -ts -s-2000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./fpgrowth - find frequent item sets with the fpgrowth algorithm
version 6.17 (2017.05.30)        (c) 2004-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
writing <null> ... [851034 set(s)] done [0.10s].
0 secs  209039 microsecs


<span style="color:green">**Answer:** </span>
| Algorithm | Time    |
| --------- | ------- |
| Apriori   | 29.12 s   |
| ECLAT     | 0.41 s |
| FP-Growth | 0.20 s |

Fastest → FP-Growth

Slowest → Apriori

FP-Growth is fastest because it avoids candidate generation and uses an FP-tree. ECLAT is also efficient using TID intersections, while Apriori is slow due to repeated database scans and large candidate generation.



<span style="color:red">**Question 2f:** </span> For a minsup = 4000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [48]:
import datetime
start = datetime.datetime.now()
!./apriori -ts -s-4000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.03s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [2.04s].
writing <null> ... [29501 set(s)] done [0.00s].
2 secs  227171 microsecs


In [49]:
import datetime
start = datetime.datetime.now()
!./eclat -ts -s-4000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./eclat - find frequent item sets with the eclat algorithm
version 5.20 (2017.05.30)        (c) 2002-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
writing <null> ... [29501 set(s)] done [0.04s].
0 secs  108012 microsecs


In [50]:
import datetime
start = datetime.datetime.now()
!./fpgrowth -ts -s-4000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./fpgrowth - find frequent item sets with the fpgrowth algorithm
version 6.17 (2017.05.30)        (c) 2004-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
writing <null> ... [29501 set(s)] done [0.02s].
0 secs  105063 microsecs


<span style="color:green">**Answer:** </span>
| Algorithm | Time    |
| --------- | ------- |
| Apriori   | 2.22 s   |
| ECLAT     | 0.108 s |
| FP-Growth | 0.105 s |

Fastest → FP-Growth (but only slightly faster than ECLAT)

Slowest → Apriori

FP-Growth is the fastest (0.105 s), ECLAT is very close (0.108 s), and Apriori is the slowest (2.23 s). FP-Growth is fast because it avoids candidate generation by mining an FP-tree, and ECLAT is fast because it uses TID-list intersections. Apriori is slower because it still generates candidates level-by-level and checks many subsets, even though the higher minsup reduces the search space.



<span style="color:red">**Question 2g:** </span>  For a minsup = 6000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [60]:
import datetime
start = datetime.datetime.now()
!./apriori -ts -s-6000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.01s].
building transaction tree ... [6478 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 done [0.06s].
writing <null> ... [2254 set(s)] done [0.00s].
0 secs  106089 microsecs


In [61]:
import datetime
start = datetime.datetime.now()
!./eclat -ts -s-6000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./eclat - find frequent item sets with the eclat algorithm
version 5.20 (2017.05.30)        (c) 2002-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.00s].
writing <null> ... [2254 set(s)] done [0.00s].
0 secs  108758 microsecs


In [62]:
import datetime
start = datetime.datetime.now()
!./fpgrowth -ts -s-6000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./fpgrowth - find frequent item sets with the fpgrowth algorithm
version 6.17 (2017.05.30)        (c) 2004-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.00s].
writing <null> ... [2254 set(s)] done [0.00s].
0 secs  105510 microsecs


<span style="color:green">**Answer:** </span>
| Algorithm | Time    |
| --------- | ------- |
| Apriori   | 0.106 s |
| ECLAT     | 0.108 s |
| FP-Growth | 0.105 s |

FP-Growth is slightly fastest

All algorithms perform similarly

FP-Growth is the fastest (0.105 s), but all three algorithms have very similar execution times. ECLAT is slightly the slowest (0.108 s).

This happens because at high minsup (6000), the number of frequent items and itemsets is very small, so the search space is As a re reduced.sult, all algorithms run very quickly and their performance becomes similar.




<span style="color:red">**Question 2h:** </span> Fill the following table based on execution times computed in __2e__, __2f__, and __2g__. State your observations on the relative computational efficiency at different support thresholds. Based on your knowledge of these algorithms, provide the reasons behind your observations.

|   Algorithm                |minsup=2000         |minsup=4000         |minsup=6000         |
|----------------------------|--------------------|--------------------|--------------------|    
|Apriori |  29.12 s  | 	2.22 s |  0.106 s                   |
|Eclat |  0.41 s  | 0.108 s |  0.108 s  |
|FPGrowth | 0.20 | 0.105 s |  0.105 s  |


<span style="color:green">**Answer:** </span>

Observations:
At minsup=2000, FP-Growth and ECLAT are much faster than Apriori. As minsup increases to 4000 and 6000, all methods become faster because the number of frequent items/itemsets shrinks, and at minsup=6000 the runtimes are almost the same (differences are only microseconds). Overall, FP-Growth is the fastest (or tied) across thresholds, and Apriori benefits the most from increasing minsup but is still the slowest at low support.

Reasons:
FP-Growth is fast because it avoids candidate generation and mines a compressed FP-tree. ECLAT is fast because it uses a vertical representation and counts support via efficient TID-list intersections. Apriori is slow at low minsup because it generates many candidates level-by-level and performs many subset checks; when minsup is higher, far fewer candidates survive, so Apriori speeds up a lot.

### 3. Discovering frequent subsequences and substrings

Assume that roads in a Cincinnati are assigned numbers. Participants are enrolled in a transportation study and for every trip they make using their car, the sequence of roads taken are recorded. Trips that involves freeways are excluded. This data is in the file <span style="color:blue">road_seq_data.dat</span>.

<span style="color:red">**Question 3a:** </span>  What 'type' of sequence mining will you perform to determine frequently taken 'paths'? Paths are sequences of roads traveresed consecutively in the same order.

<span style="color:green">**Answer:** </span> Substring mining, because paths require roads to appear consecutively in the same order, not just in order with gaps.

<span style="color:red">**Question 3b:** </span> How many sequences are there in this sequence database?

In [65]:
!wc -l road_seq_data.dat

1000 road_seq_data.dat


<span style="color:green">**Answer:** </span> Number of sequences in the data file road_seq_data are 1000

<span style="color:red">**Question 3c:** </span> What is the size of the alphabet in this sequence database?

In [66]:
!awk -- '{for (i = 1; i <= NF; i++) wc[$i] += 1}; END {print length(wc)}' road_seq_data.dat

1283


<span style="color:green">**Answer:** </span> Size of the alphabet in the data file road_seq_data is 1283

<span style="color:red">**Question 3d:** </span> What are the total number of possible subsequences of length 2 in this dataset?

Formual: (Σ)<sup>l<sup>

In [70]:
1283**2

1646089

<span style="color:green">**Answer:** The total number of possible subsequences of length 2 is 1283² = 1646089.

<span style="color:red">**Question 3e:** </span> What are the total number of possible substrings of length 2 in this dataset?

Formual: (Σ)<sup>l<sup>

In [72]:
1283**2

1646089

<span style="color:green">**Answer:** The total number of possible substrings of length 2 is |Σ|^2 = 1283^2 = 1646089.

<span style="color:red">**Question 3f:** </span> Discover frequent __subsequences__ with minsup = 10 and report the number of subsequences discovered.

In [92]:
!chmod u+x prefixspan

In [93]:
!./prefixspan

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei


Usage: prefixspan [OPTION]... INFILE

       where [OPTION]...  is a list of zero or more optional arguments
             INFILE(s)    is the name of the input transaction database

Additional arguments (at most one input file may be specified):
       -min_sup [minimum support]
       -max_pat [maximum pattern]



In [94]:
!./prefixspan -min_sup 10 road_seq_data.dat | sed -n 'p;n' > road_seq_data_minsup_10

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei



In [95]:
!wc -l road_seq_data_minsup_10

4589 road_seq_data_minsup_10


<span style="color:green">**Answer:** </span> The number of frequent subsequences discovered with minsup = 10 is 4589.

<span style="color:red">**Question 3g:** </span>  Discover frequent __substrings__ with minsup = 10 and report the number of substrings discovered.

In [96]:
!chmod u+x seqwog

In [97]:
!./seqwog -ts -s-10 road_seq_data.dat road_seq_data_substring_result

./seqwog - find frequent sequences without gaps
version 3.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
reading road_seq_data.dat ... [1283 item(s), 1000 transaction(s)] done [0.00s].
recoding items ... [1283 item(s)] done [0.00s].
reducing and triming transactions ... [844/1000 transaction(s)] done [0.00s].
writing road_seq_data_substring_result ... [613 sequence(s)] done [0.00s].


In [98]:
!wc -l road_seq_data_substring_result

613 road_seq_data_substring_result


<span style="color:green">**Answer:** </span> The number of frequent substrings discovered with minsup = 10 is 613.

<span style="color:red">**Question 3h:** </span> Explain the difference in the number of frequent subsequences and substrings found in __3f__ and __3g__ above.

<span style="color:green">**Answer:** </span> There are many more frequent subsequences (4589) than substrings (613) because subsequences allow gaps between items, while substrings must be consecutive. This makes subsequences less restrictive, so more patterns satisfy the minimum support.

In [99]:
4589 - 613

3976